akka-stream-tests/src/test/scala/akka/stream/scaladsl/AttributesSpec.scala

```diff
@@ -45,28 +104,167 @@ class AttributesSpec extends StreamSpec {
+          .futureValue
 
-    "give access to first attribute" in {
-      attributes.getFirst[Name] should ===(Some(Attributes.Name("a")))
+      // most specific
+      attributes.get[Name] should contain(Name("new-name"))
+
+      // least specific
+      attributes.getFirst[Name] should contain(Name("new-name"))
     }
 
-    "give access to attribute byt type" in {
-      attributes.get[Name] should ===(Some(Attributes.Name("b")))
+    "make the attributes on Source.fromGraph source behave the same as the stage itself" in {
+      val attributes =
+        Source.fromGraph(new AttributesSource(Attributes.name("original-name")))
+          .withAttributes(Attributes.name("replaced")) // this actually replaces now
+          .toMat(Sink.head)(Keep.right).withAttributes(Attributes.name("whole-graph"))
+          .run()
+          .futureValue
+
+      // most specific
+      attributes.get[Name] should contain(Name("replaced"))
+
+      // least specific
+      attributes.getFirst[Name] should contain(Name("whole-graph"))
+    }
+
+    "make the attributes on Flow.fromGraph source behave the same as the stage itself" in {
+      val attributes =
+        Source.maybe
+          .viaMat(
+            Flow.fromGraph(new AttributesFlow(Attributes.name("original-name")))
+              .withAttributes(Attributes.name("replaced")) // this actually replaces now
+          )(Keep.right)
+          .withAttributes(Attributes.name("source-flow"))
+          .toMat(Sink.ignore)(Keep.left)
+          .withAttributes(Attributes.name("whole-graph"))
+          .run()
+
+      // most specific
+      attributes.get[Name] should contain(Name("replaced"))
+
+      // least specific
+      attributes.getFirst[Name] should contain(Name("whole-graph"))
+    }
+
+    "make the attributes on Sink.fromGraph source behave the same as the stage itself" in {
+      val attributes =
+        Source.maybe.toMat(
+          Sink.fromGraph(new AttributesSink(Attributes.name("original-name")))
+            .withAttributes(Attributes.name("replaced")) // this actually replaces now
+        )(Keep.right)
+          .withAttributes(Attributes.name("whole-graph"))
+          .run()
+
+      // most specific
+      attributes.get[Name] should contain(Name("replaced"))
+
+      // least specific
+      attributes.getFirst[Name] should contain(Name("whole-graph"))
+    }
+
+    "use the initial attributes for dispatcher" in {
+      val dispatcher =
+        Source.fromGraph(new ThreadNameSnitchingStage("my-dispatcher"))
+          .runWith(Sink.head)
+          .futureValue
+
+      dispatcher should startWith("AttributesSpec-my-dispatcher")
+    }
+
+    "use the most specific dispatcher when specified directly around the graph stage" in {
+      val dispatcher =
+        Source.fromGraph(
+          // directly on stage
+          new ThreadNameSnitchingStage("akka.stream.default-blocking-io-dispatcher")
+            .addAttributes(ActorAttributes.dispatcher("my-dispatcher")))
```

Comment:
is that the same as
```
Source.fromGraph(new ThreadNameSnitchingStage("akka.stream.default-blocking-io-dispatcher"))
    .addAttributes(ActorAttributes.dispatcher("my-dispatcher"))
```
?

Reply:
With my fix it is, before it wasn't because of the wrapping.